core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -249,6 +285,137 @@ public synchronized void updateAllocationIdsFromMaster(
     }
 
     /**
+     * Get the primary context for the shard. This includes the state of the global checkpoint tracker.
+     *
+     * @return the primary context
+     */
+    synchronized Releasable primaryContext(final Consumer<PrimaryContext> consumer) {
+        if (sealed) {
+            throw new IllegalStateException("global checkpoint tracker is sealed");
+        }
+        sealed = true;
+        final ObjectLongMap<String> inSyncLocalCheckpoints = new ObjectLongHashMap<>(this.inSyncLocalCheckpoints);
+        final ObjectLongMap<String> trackingLocalCheckpoints = new ObjectLongHashMap<>(this.trackingLocalCheckpoints);
+        try {
+            consumer.accept(new PrimaryContext(appliedClusterStateVersion, inSyncLocalCheckpoints, trackingLocalCheckpoints));
+        } catch (final Exception e) {
+            sealed = false;
+            throw e;
+        }
+        return () -> {
```

Comment:
Well, it was [your suggestion](https://github.com/elastic/elasticsearch/pull/25122#discussion_r122955694) to return a `Releasable`. Yet, I will change this. 😇